Public/Src/Tools/Execution.Analyzer/Analyzers.Core/XLGPlusPlus/StaticGraph.proto

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+syntax = "proto3";
+
+import "EnumsAndHelperStructs.proto";
+
+package BuildXL.Execution.Analyzer;
+
+option csharp_namespace = "BuildXL.Execution.Analyzer.Xldb";
+
+//
+//  Generic Static Graph Info (key, etc)
+//
+
+// CachedGraphQuery is the key for the top level data structures stored in the cached graph
+message CachedGraphQuery{
+    bool PipTable = 1;
+
+    bool PipGraph = 2;  
+}
+
+// Pip Query is 2 levels deep. SemistableHash -> PipId -> PipValues
+// For the pips with SemiStableHash = 0, we will only have PipId -> Values
+message PipQuerySemiStableHash{
+    // Primary key, unique except for the pips that have a SemiStableHash = 0
+    int64 SemiStableHash = 1;
+}
+
+message PipValueSemiStableHash{
+    uint32 PipId = 1;
+}
+
+message PipQueryPipId{
+    // Always unique within a build (makes the key unique), and same as the node id in graph
+    uint32 PipId = 1;
+
+    // We include the PipType in the key to know which type of pip is stored in the value
+    // and how to appropriately parse it (since Protobuf has no inheritance)
+    PipType PipType = 2;
+}
+
+//
+//  Map Messages -> Protobuf does not allow for float, double, bytes, or message types to be 
+//  used as keys for maps, so the alternative is to make a "primitive" map type message as follows:
+//  
+//  message primitiveMap{ key; value; }
+//  message wraper { repeated primitiveMap; }
+//
+
+message FileArtifactMap{
+    FileArtifact Artifact = 1;
+
+    uint32 Value = 2;
+}
+
+message DirectoryArtifactMap{
+    DirectoryArtifact Artifact = 1;
+
+    uint32 Value = 2;
+}
+
+message PipIdSet{
+    repeated uint32 PipIds = 1;
+}
+
+message NodeValuesKey{
+    int32 FullSymbol = 1;
+
+    int32 QualifierId = 2;
+
+    AbsolutePath Path = 3;
+}
+
+message NodeValuesMap{
+    NodeValuesKey Key = 1;
+
+    uint32 Value = 2;
+}
+
+message PipDeserializationContext{
+    PipQueryContext Key = 1;
+
+    int32 Value = 2;
+}
+
+message PipGraphStaticFingerprints{
+    map<uint32, ContentFingerprint> PipStaticFingerprints = 1;
+}
+
+//
+//  Graph Data
+//
+
+message PipTable{

```

Comment:
>PipTable [](start = 8, length = 8)

Pip table is just the list of pips. There should be no need to have a separate data structure for this.

Reply:
This just contains metadata about the piptable such as the number of reads/writes, the different stablekeys, and more. Perhaps I can rename it to PipTableMetaData or something similar

---
In reply to: [314336956](https://github.com/microsoft/BuildXL/pull/724#discussion_r314336956) [](ancestors = 314336956)